python/simulation.py

```diff
@@ -4425,6 +4425,41 @@ def _stop(sim):
 
     return _stop
 
+def stop_when_dft_decayed(decay_by, minimum_run_time=0, maximum_run_time=None):
+    """
+    Return a `condition` function, suitable for passing to `sim.run()` as the `until`
+    or `until_after_sources` parameter, that checks all of the Simulation's dft objects
+    every `dt` timesteps, and stops the simulation once all the components and frequencies
+    of *every* dft object have decayed by at least `decay_by`. The routine calculates the optimal
+    `dt` to use based on the frequency content in the DFT monitors.
+
+    Optionally the user can specify a minimum run time (`minimum_run_time`) or a maximum
+    run time (`maximum_run_time`).
+    """
+    # Record data in closure so that we can persitently edit
+    closure = {'previous_fields':1, 't0':0, 'dt':0}
+    def _stop(_sim):
+        if _sim.fields.t == 0:
+            closure['dt'] = max(int(1/_sim.fields.dft_maxfreq()/_sim.fields.dt),int(_sim.fields.min_decimation()))
+        if _sim.fields.t <= closure['dt'] + closure['t0']:
+            return False
+        elif maximum_run_time and _sim.round_time() > maximum_run_time:
+            return True
```

Comment:
Shouldn't the second if-branch which checks if the `maximum_run_time` has been exceeded come *before* the first if-branch which checks that a time interval of `dt` has elapsed?

Reply:
Fixed.